src/audio/fork_trigger.c

```diff
@@ -0,0 +1,280 @@
+	struct list_item *blist;
+	int count = 0;
+
+	/* count sources with state == status */
+	list_for_item(blist, &dev->bsource_list) {
+		source = container_of(blist, struct comp_buffer, sink_list);
+		comp_info(source->source, "source state %u", source->source->state);
+		if (source->source->state == status)
+			count++;
+	}
+
+	return count;
+}
+
+static int sink_status_count(struct comp_dev *dev, unsigned int status)
+{
+	struct comp_buffer *sink;
+	struct list_item *blist;
+	int count = 0;
+
+	/* count sinks with state == status */
+	list_for_item(blist, &dev->bsink_list) {
+		sink = container_of(blist, struct comp_buffer, source_list);
+		comp_info(sink->sink, "sink state %u", sink->sink->state);
+		if (sink->sink->state == status)
+			count++;
+	}
+
+	return count;
+}
+
+static unsigned int fork_status_count(struct comp_dev *dev,
+				      unsigned int status)
+{
+	switch (dev_comp_type(dev)) {
+	case SOF_COMP_MIXER:
+	case SOF_COMP_MUX:
+		return source_status_count(dev, status);
+	case SOF_COMP_DEMUX:
+		return sink_status_count(dev, status);
+	default:
+		return 0;
+	}
+}
+
+static int fork_handle_stop(struct comp_dev *dev)
+{
+	switch (dev->state) {
+	case COMP_STATE_ACTIVE:
+		/* from flow: extended for multiple inputs */
+		if (fork_status_count(dev, COMP_STATE_ACTIVE))
+			return PPL_STATUS_PATH_STOP;
+
+		if (fork_status_count(dev, COMP_STATE_PAUSED)) {
+			/* only one active pipeline, but other pipelines are paused */
+			dev->cmd_override = COMP_TRIGGER_PAUSE;
+			dev->state = COMP_STATE_PAUSED;
+			return 0;
+		}
+
+		dev->state = COMP_STATE_PREPARE;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int fork_handle_start(struct comp_dev *dev)
+{
+	switch (dev->state) {
+	case COMP_STATE_PRE_ACTIVE:
+		if (fork_status_count(dev, COMP_STATE_PAUSED)) {
+			/* we were in PAUSED when a PRE_START arrived */
+			dev->cmd_override = COMP_TRIGGER_RELEASE;
+		}
+
+		/* from flow */
+		dev->state = COMP_STATE_ACTIVE;
+		return 0;
+	case COMP_STATE_ACTIVE:
+		return PPL_STATUS_PATH_STOP;
+	}
+
+	return -EINVAL;
+}
+
+static int fork_handle_pause(struct comp_dev *dev)
+{
+	switch (dev->state) {
+	case COMP_STATE_ACTIVE:
+		/* from flow: extended for multiple inputs */
+		if (fork_status_count(dev, COMP_STATE_ACTIVE) ||
+		    fork_status_count(dev, COMP_STATE_PAUSED) > 1)
+			/* Keep the current state */
+			return PPL_STATUS_PATH_STOP;
+
+		dev->state = COMP_STATE_PAUSED;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int fork_handle_release(struct comp_dev *dev)
+{
+	switch (dev->state) {
+	case COMP_STATE_ACTIVE:
+		/*
+		 * the mixer and its pipeline didn't suspend - they have
+		 * additional active source pipelines
+		 */
+		return PPL_STATUS_PATH_STOP;
+	case COMP_STATE_PRE_ACTIVE:
+		/* from flow */
+		dev->state = COMP_STATE_ACTIVE;
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int fork_handle_suspend(struct comp_dev *dev)
+{
+	return -EINVAL;
+}
+
+static int fork_handle_resume(struct comp_dev *dev)
+{
+	return -EINVAL;
+}
```

Comment:
I dont think we use suspend or resume at all (but pls check), this was taken from ALSA but I dont think every used. Could be removed to simplify code.